analytical_engine/apps/simple_path/all_simple_paths.h

```diff
@@ -0,0 +1,324 @@
+      vid_t msg2 = std::get<0>(msg).second;
+      // msg[2] == true means init frag_vertex_num, msg[1] == false means new
+      // vertex to deal with, msg[1] == true means update edge_map.
+      if (std::get<2>(msg) == true) {
+        init_counter++;
+        ctx.frag_vertex_num[gid] = msg2;
+        if (init_counter == static_cast<int>(frag.fnum()) - 1) {
+          vertex_t source;
+          frag.GetInnerVertex(ctx.source_id, source);
+          vertexProcess(source, frag, ctx, messages, 0);
+          vid_t s_gid;
+          frag.Oid2Gid(ctx.source_id, s_gid);
+          ctx.visit.insert(s_gid);
+          messages.ForceContinue();
+          frag_finish_counter++;
+        }
+      } else if (std::get<1>(msg) == false) {
+        if (ctx.visit.count(gid) == 0) {
+          ctx.visit.insert(gid);
+          ctx.next_level_inner.push(std::make_pair(gid, msg2));
+        }
+      } else if (ctx.source_flag == true && std::get<1>(msg) == true) {
+        int a = findVertexGlobalIndex(ctx, gid);
+        int b = findVertexGlobalIndex(ctx, msg2);
+        ctx.edge_map[a].push_back(b);
+      }
+    }
+    ctx.curr_level_inner.swap(ctx.next_level_inner);
+    while (!ctx.curr_level_inner.empty()) {
+      vid_t gid = ctx.curr_level_inner.front().first;
+      int depth = ctx.curr_level_inner.front().second;
+      ctx.curr_level_inner.pop();
+      vertex_t v;
+      frag.Gid2Vertex(gid, v);
+      if (depth <= ctx.cutoff) {
+        bool ret = vertexProcess(v, frag, ctx, messages, depth);
+        if (ret == true)
+          frag_finish_counter++;
+      }
+    }
+    if (!ctx.next_level_inner.empty() || frag_finish_counter > 0)
+      messages.ForceContinue();
+    Sum(frag_finish_counter, ctx.frag_finish_counter);
+    if (ctx.frag_finish_counter == 0 && frag.fid() == ctx.soucre_fid) {
+      writeToCtx(frag, ctx);
+    }
+  }
+
+ private:
+  bool vertexProcess(vertex_t v, const fragment_t& frag, context_t& ctx,
+                     message_manager_t& messages, int depth) {
+    bool ret = false;
+    std::set<vid_t> pvisit;
+    vid_t gid = frag.Vertex2Gid(v);
+    auto oes = frag.GetOutgoingAdjList(v);
+    for (auto& e : oes) {
+      vertex_t u = e.get_neighbor();
+      vid_t u_gid = frag.Vertex2Gid(u);
+      if (pvisit.count(u_gid) == 0) {
+        pvisit.insert(u_gid);
+        if (ctx.source_flag == false) {
+          std::tuple<std::pair<vid_t, vid_t>, bool, bool> msg =
+              std::make_tuple(std::make_pair(gid, u_gid), true, false);
+          messages.SendToFragment(ctx.soucre_fid, msg);
+          ret = true;
+        } else {
+          int a = findVertexGlobalIndex(ctx, gid);
+          int b = findVertexGlobalIndex(ctx, u_gid);
+          ctx.edge_map[a].push_back(b);
+        }
+        if (!frag.IsOuterVertex(u)) {
+          if (ctx.visit.count(u_gid) == 0) {
+            ctx.visit.insert(u_gid);
+            ctx.next_level_inner.push(std::make_pair(u_gid, depth + 1));
+          }
```

Comment:
if the `ctx.visit` is only store inner vertices of fragment , better use a `vertex_array_t<bool> visited` to replace it.
ref https://github.com/alibaba/GraphScope/blob/db3b5e2a499e1c46cfaf411ad91314750b32216e/analytical_engine/apps/bfs/bfs_generic_context.h#L82